enterprise/server/scheduling/priority_task_scheduler/priority_task_scheduler.go

```diff
@@ -474,6 +478,58 @@ func (q *PriorityTaskScheduler) canFitTask(res *scpb.EnqueueTaskReservationReque
 	return true
 }
 
+// This function peeks at the front of the queue and checks to see if the task
+// still exists--if it doesn't, then we know the task was picked up *and
+// finished* by another worker, so we can safely prune it from the local queue,
+// allowing our queue size to go down even when we aren't picking up new tasks.
+//
+// During sudden bursts of large tasks, the task queues on individual executors
+// can get pretty long.  If we are autoscaling according to task queue length,
+// then doubling the number of executors might completely empty out the "real"
+// task queue, but if the original executor pool is fully stuck on large tasks,
+// the average task queue length will only be cut in half!  If that new length
+// is still over our autoscaling threshold, we'll scale up more even though we
+// don't need to.  Trimming makes this case less likely.
+//
+// This function returns true if a task was successfully dequeued.
+func (q *PriorityTaskScheduler) trimQueue() bool {
+	q.mu.Lock()
+	if q.shuttingDown || q.q.Len() == 0 {
+		return false
+	}
+	nextTask := q.q.Peek()
+	if nextTask == nil {
+		return false
+	}
+	// Hand the lock back in case this RPC takes a while.
+	q.mu.Unlock()
+
+	ctx := log.EnrichContext(q.rootContext, log.ExecutionIDKey, nextTask.GetTaskId())
+	ctx = tracing.ExtractProtoTraceMetadata(ctx, nextTask.GetTraceMetadata())
+	resp, err := q.env.GetSchedulerClient().TaskExists(ctx, &scpb.TaskExistsRequest{TaskId: nextTask.GetTaskId()})
+	if err != nil {
+		log.Infof("Failed to check if task exists: %s", err)
+		return false
+	}
+	if resp.GetExists() {
+		// This task hasn't been finished by another executor.  Don't prune.
+		return false
+	}
+
+	// Now that we know the task is gone, make sure it's still at the front of the queue.
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	if nextTask != q.q.Peek() {
+		return false
+	}
+
+	// Queue hasn't changed--since the task is gone, that means we can safely remove it.
+	q.q.Dequeue()
```

Comment:
uber nit: might be nice to add a little safety check here (I could see us accidentally making the queue non-deterministic later...)

```suggestion
	if t := q.q.Dequeue(); nextTask != t {
		alert.UnexpectedEvent("nondeterministic_dequeue", "Dequeue() returned a different value than what Peek() returned")
		return false
	}
```